node/src/components/consensus/highway_core/state.rs

```diff
@@ -361,54 +363,47 @@ impl<C: Context> State<C> {
     pub(crate) fn validate_vote(&self, swvote: &SignedWireVote<C>) -> Result<(), VoteError> {
         let wvote = &swvote.wire_vote;
         let creator = wvote.creator;
-        wvote.panorama.validate(self)?;
-        let mut justifications = wvote.panorama.iter_correct();
-        if !justifications.all(|vh| self.vote(vh).timestamp <= wvote.timestamp) {
+        let panorama = &wvote.panorama;
+        let timestamp = wvote.timestamp;
+        panorama.validate(self)?;
+        if panorama.iter_correct(self).any(|v| v.timestamp > timestamp) {
             return Err(VoteError::Timestamps);
         }
-        match wvote.panorama.get(creator) {
-            Observation::Faulty => {
-                error!("Vote from faulty validator should be rejected in `pre_validate_vote`.");
-                return Err(VoteError::FaultyCreator); // Should be unreachable.
+        if wvote.seq_number != panorama.next_seq_num(self, creator) {
+            return Err(VoteError::SequenceNumber);
+        }
+        let opt_prev_vote = panorama[creator].correct().map(|vh| self.vote(vh));
+        let prev_round_exp = opt_prev_vote.map_or(self.params.init_round_exp(), |v| v.round_exp);
+        let max_re = prev_round_exp.max(wvote.round_exp);
+        // The round exponent must only change one step at a time.
+        if prev_round_exp + 1 < max_re || wvote.round_exp + 1 < max_re {
+            return Err(VoteError::RoundLength);
+        }
+        if let Some(prev_vote) = opt_prev_vote {
+            if prev_round_exp != wvote.round_exp
+                && prev_vote.timestamp >> max_re == timestamp >> max_re
```

Comment:
What is this test `prev_vote.timestamp >> max_re == timestamp >> max_re` checking?

Reply:
That if the round exponent changed, there is a round boundary between the votes. I'll add a comment.